### PR TITLE
Revoke all authorized applications on password reset

### LINKED ISSUE #20431: password reset doesn't reset authorized apps
### Steps to reproduce the problem

1. log into mastodon using a mobile application (i.e. Fedilab)
2. log into mastodon using a webbrowser and the same account and navigate to `/auth/edit`
3. update the password
4. check the mobile application and load new articles, it still works although it shouldn't

### Expected behaviour

The authorized app tokens should be revoked and the mobile application should require the new password.

### Actual behaviour

The mobile application still works without the need to reset the password.

### Detailed description

Yesterday evening, I've resetted my password on an instance running on v3.5.3.
I checked my smartphone, because I assumed to enter the new password. However, I could still load new articles, so I assumed it might take time.
Today, I've checked my smartphone again and could still load new articles. I checked `/oauth/authorized_applications` and the app was still authorized.

### Specifications

Mastodon v3.5.3

### PR DESCRIPTION
**Pull request by FrancisMurillo**
Clear all user sessions after a successful password reset if this is the desired behavior.  I noticed that `User::reset_password!` is not really being invoked by the password reset flow so I am not sure if this is intentional as the tests are working as well. I would also like to add a controller test for this behavior but would need some guidance.

For change password, the behavior is the same and may not clear the sessions except the current. So this might also be considered.

Steps I took when I tested with my instance:
1. Login with the Mastodon iOS client
2. Login with the Web interface
3. Complete reset password with another browser
4. Refreshing on iOS client should return token revoked message
5. Refreshing on Web interface should return to start page

### CODE REVIEW

**Review by mrbrunelli — APPROVED** [review]

File: app/models/user.rb [hunk 1]
```diff
@@ -373,6 +373,15 @@ def reset_password(new_password, new_password_confirmation)
     super
   end
 
+  def clear_sessions!
```
[the remark this answers is not in the document]
Comment (FrancisMurillo) [reply app/models/user.rb:376]: Any suggestions for a clearer name?

[the remark this answers is not in the document]
Comment (ineffyble) [reply app/models/user.rb:376]: Hmm, maybe `revoke_access`?

[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/models/user.rb]: `revoke_tokens` would be even more explicit, I think

[the remark this answers is not in the document]
Comment (ineffyble) [reply app/models/user.rb]: I considered that, but it's also revoking access grants and deleting push subscriptions.

Maybe that's implementation detail.

**Comment by ClearlyClaire** [comment]
Thank you for your contribution! I don't think it's addressing #20431 as this issue was explicitly about `/auth/edit`, which is different from the password reset flow, but this is not necessarily a problem (apart from incorrectly marking #20431 as resolved).

I am not completely sure we want that behavior, especially since current apps handle revoked tokens quite poorly (see https://github.com/mastodon/mastodon/issues/20431#issuecomment-1323511493) but I understand the reasoning, and I think revoking tokens in the password reset flow as addressed in this PR would cause less issue than doing it in the `/auth/edit` flow.

> I noticed that `User::reset_password!` is not really being invoked by the password reset flow so I am not sure if this is intentional as the tests are working as well.

The password reset is taken care of by `Devise::PasswordsController#update` which is called through the `super` call. The block is then executed right after that. I think not calling `reset_password!` here is intentional.

> I would also like to add a controller test for this behavior but would need some guidance.

That would be great! This would be a new `describe` block for `POST #update` in `spec/controllers/auth/passwords_controller_spec.rb`. You could read the tests for `#reset_password!` in `spec/models/user_spec.rb` for inspiration.

**Comment by FrancisMurillo** [comment]
I started with the password reset to understand if this behavior is acceptable which I was hoping then to reuse for change password without removing the current session that is. If this behavior is currently not desired, I will remove marking the issue resolved. For now, will just add the controller test.

File: app/models/user.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrancisMurillo) [reply app/models/user.rb]: If I refactor this method to also delete `session_activations` if it was used for change password, would the original name be more helpful?

File: spec/controllers/auth/passwords_controller_spec.rb [hunk 2]
```diff
@@ -35,4 +35,65 @@
...
+    end
+
+    context 'with invalid reset_password_token' do
+      before do
+        post :update, params: { user: { reset_password_token: 'some_invalid_value' } }
+      end
+
+      it 'returns http success' do
+        expect(response).to have_http_status(200)
+      end
```
Comment (ClearlyClaire) [inline spec/controllers/auth/passwords_controller_spec.rb:89]: Is that actually the expected behavior?

Comment (FrancisMurillo) [reply spec/controllers/auth/passwords_controller_spec.rb]: I see your point. So I changed the test expectation to redirection on success and render a error template on failure which might be better.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by wojtekxtx** [comment]
@ClearlyClaire you know what would be great? To close down this stupid wannabe serious "platform" (mastodon) and move on.

**Comment by ineffyble** [comment]
@FrancisMurillo your commits are showing up as Unverified, as you've requested GitHub to require GPG signing for your account. Can you look into this?

**Comment by FrancisMurillo** [comment]
@ineffyble Was able to finally recover my keys and gpg sign the commits so they are now marked as verified.

**Pull request merged**